File: Design-Patterns-Using-Go/Observer/ObserverPattern.go
package main

import "fmt"

// Subject interface
type Subject interface {
	Attach(observer Observer)
	Detach(observer Observer)
	Notify()
}

// ConcreteSubject struct
type ConcreteSubject struct {
	state     string
	observers []Observer
}

func (s *ConcreteSubject) Attach(observer Observer) {
	observer.SetSubject(s)
	s.observers = append(s.observers, observer)
}

func (s *ConcreteSubject) Detach(observer Observer) {
	for i, obs := range s.observers {
		if obs == observer {
			observer.SetSubject(nil)
			s.observers = append(s.observers[:i], s.observers[i+1:]...)
			break
		}
	}
}

func (s *ConcreteSubject) Notify() {
	for _, observer := range s.observers {
		observer.Update()
	}
}

func (s *ConcreteSubject) GetState() string {
	return s.state
}

func (s *ConcreteSubject) SetState(state string) {
	s.state = state
	s.Notify()
}

// Observer interface
type Observer interface {
	Update()
	SetSubject(subject Subject)
}

// ConcreteObserver struct
type ConcreteObserver struct {
	subject Subject
}

func (o *ConcreteObserver) Update() {
	state := o.subject.(*ConcreteSubject).GetState()
	fmt.Printf("%s notified to Observer\n", state)
}

func (o *ConcreteObserver) SetSubject(subject Subject) {
	o.subject = subject
}

// Client code
func main() {
	subject := &ConcreteSubject{}
	observer1 := &ConcreteObserver{}
	observer2 := &ConcreteObserver{}

	subject.Attach(observer1)
	subject.Attach(observer2)

	subject.SetState("First state")
	subject.SetState("Second state")
}

/*
First state notified to Observer
First state notified to Observer
Second state notified to Observer
Second state notified to Observer
*/

File: Design-Patterns-Using-Cpp/Observer/ObserverPattern.cpp
#include <iostream>
#include <vector>
#include <algorithm>

class Subject;

class Observer {
public:
    virtual void update(const std::string& newState) = 0;
    virtual void setSubject(Subject* subject) = 0;
};

class Subject {
private:
    std::vector<Observer*> observers;

public:
    void attach(Observer* observer) {
        observer->setSubject(this);
        observers.push_back(observer);
    }

    void detach(Observer* observer) {
        observer->setSubject(nullptr);
        auto it = std::find(observers.begin(), observers.end(), observer);
        observers.erase(it);
    }

    void notifyObservers(const std::string& newState) {
        for (Observer* observer : observers) {
            observer->update(newState);
        }
    }
};

class ConcreteSubject : public Subject {
private:
    std::string state;

public:
    const std::string& getState() const {
        return state;
    }

    void setState(const std::string& newState) {
        state = newState;
        notifyObservers(newState);
    }
};

class ConcreteObserver1 : public Observer {
private:
    Subject* subject;

public:
    ConcreteObserver1(Subject* subject) : subject(subject) {
        subject->attach(this);
    }

    void update(const std::string& newState) override {
        std::cout << newState << " notified to Observer1" << std::endl;
    }

    void setSubject(Subject* newSubject) override {
        subject = newSubject;
    }
};

class ConcreteObserver2 : public Observer {
private:
    Subject* subject;

public:
    ConcreteObserver2(Subject* subject) : subject(subject) {
        subject->attach(this);
    }

    void update(const std::string& newState) override {
        std::cout << newState << " notified to Observer2" << std::endl;
    }

    void setSubject(Subject* newSubject) override {
        subject = newSubject;
    }
};

// Client code
int main() {
    ConcreteSubject subject;
    ConcreteObserver1 observer1(&subject);
    ConcreteObserver2 observer2(&subject);

    subject.setState("First state");
    subject.setState("Second state");

    return 0;
}

/*
First state notified to Observer1
First state notified to Observer2
Second state notified to Observer1
Second state notified to Observer2
*/